lecture_notes/2017-02-14.md

```diff
@@ -0,0 +1,362 @@
+because:
+- with it, Gentzen could prove that the sequent calculus is consistent
+  (by showing there is no way to prove a contradiction without using cuts)
+- cut elimination implies a _subformula property_ --- that one can build a
+  larger proof from smaller proofs, just as one can build a large program by
+  composing functions
+
+#### Linear Logic
+
+Linear logic is similar to the sequent calculus.
+
+Biggest difference is the absence of structural rules:
+- no _contraction_, cannot duplicate assumptions in the context
+- no _weakening_, cannot discard assumptions in the context
+
+When proving `Γ ⊢ A`, every fact in `Γ` must be used _exactly once_.
+Consequently, the sequent calculus rules for products no longer work out (would
+need to discard/duplicate propositions). Instead, there are 2 kinds of products
+in linear logic:
+- "tensor" (`A ⊗ B`) : to eliminate a tensor, must use both facts at once
+- "with" (`A & B`) : can project either conjunct, but to introduce a "with" need
+  to prove both sides simultaneously
+
+Girard also has an "of course" operator (`A!`). Contraction and weaking holds
+for "of course" propositions. If you stick to just "of course"s, you're back
+to intuitionistic logic.
+
+> M: question: is ! analogous to monads, in the sense that it introduces a
+>    controlled way of escaping to intuitionistic logic?
+> N: yep, it's a co-modality and monads are just modalities. It's a co-monad.
+
+### 1988 LaFont
+
+_The Linear Abstract Machine_
+
+> M: I remember reading that paper!
+
+Idea: build a computational model of linear logic
+
+> W: wait, can we go back to linear logic? Since we just added intuitionistic
+>    with "of course", how do we encode "bottom"?
+> N: the axiom `Γ, 0 ⊢ A`. (0 is the encoding of ⊥)
+
+> J: back to code?
+> M: hah! I remember reading this _Linear Abstract Machine_ paper and it was
+>    such a letdown when I finished
+
+The linear abstract machine is defined with a transition function from states
+to states.
+```
+ ⟨ Code , Env , Stack ⟩ → ⟨ Code' , Env' , Stack' ⟩
+```
+
+And the system has some interesting properties:
+- doesn't need a garbage collector, it's safe to deallocate a value once its used
+- can add mutation / side effects without screwing anything up
+
+> M: I did that --- added side effects --- when I first read the paper.
+>    Those "side effects" wind up being completely trivial.
+> D: yep, since you're the only one who can mutate anything, it's basically
+>    the same as making a copy
+> D: we'll see that the recent focus (post 1990) has been in bringing these
+>    desirable features to a usable programming language.
+
+> R: does the machine have "of course"
+> D: not sure ... oh yes, it does. Implemented by copying. There is a lot of
+>    copying
+
+
+### 1988 Holmstrom
+
+_Linear Functional Programming_
+
+This paper is cited, but I couldn't find a copy. Even emailed Simon Peyton Jones
+(the workshop organizer?) and he said his shelf for that conference only went
```

Comment:
No '?' needed - he was one of the workshop editors. Only way I found that out was that the citations I found were "proceedings of X workshop, editors ..."